File: server/lib/export/bundle.ts
/**
 * Assembles an export bundle's zip and uploads it. Called by the queue
 * (lib/export/queue.ts) once the underlying model artifact exists — either
 * because it already did (model tier re-export) or because the worker's
 * conversion finished (see the `export` case in lib/microservice.ts).
 */

import CONSTANTS from '@schema/constants.json'
import { db } from '@server/db'
import { modelExports } from '@server/db/schema'
import type { ExportFormat } from '@server/lib/enums'
import { bundleKey, downloadFile, exportKey, fileExists, uploadFile } from '@server/lib/storage'
import { getTaskDescriptor } from '@server/lib/tasks'
import { record } from '@server/lib/telemetry'
import { eq } from 'drizzle-orm'
import { strToU8, zipSync } from 'fflate'
import { extractPreprocessing } from './metadata'
import { renderReadme } from './readme'
import { render, templates } from './templates'

interface ExpectedJson {
  schemaVersion: 1
  inputColumn: string
  inputValue: unknown
  outputColumn: string
  outputType: string
  predictions: Record<string, number>
}

interface ResolvedSample {
  /** e.g. "input.jpg", "input.txt", "input.json" */
  filename: string
  bytes: Uint8Array
}

/**
 * Turn expected.json's raw `inputValue` (an `s3://` URI for file-backed
 * modalities, or an inline text/scalar value otherwise) into an actual file
 * to embed in the bundle under `sample/`.
 */
async function resolveSampleFile(inputValue: unknown): Promise<ResolvedSample | null> {
  if (typeof inputValue === 'string' && inputValue.startsWith('s3://')) {
    const withoutScheme = inputValue.slice('s3://'.length)
    const slashIdx = withoutScheme.indexOf('/')
    if (slashIdx === -1) return null
    const bucket = withoutScheme.slice(0, slashIdx)
    const key = withoutScheme.slice(slashIdx + 1)
    const dotIdx = key.lastIndexOf('.')
    const ext = dotIdx === -1 ? '' : key.slice(dotIdx)
    try {
      const bytes = await downloadFile(bucket, key)
      return { filename: `input${ext}`, bytes }
    } catch {
      return null
    }
  }
  if (typeof inputValue === 'string') {
    return { filename: 'input.txt', bytes: strToU8(inputValue) }
  }
  if (inputValue !== undefined && inputValue !== null) {
    return { filename: 'input.json', bytes: strToU8(JSON.stringify(inputValue, null, 2)) }
  }
  return null
}

/** Load the worker-written expected.json (if any) and its resolved sample file, rewritten to bundle-local paths. */
async function loadGoldenSample(runId: string): Promise<{ expectedJson: Uint8Array; sample: ResolvedSample } | null> {
  const exists = await fileExists(CONSTANTS.BUCKET_MODELS, `${runId}/expected.json`)
  if (!exists) return null

  const raw = await downloadFile(CONSTANTS.BUCKET_MODELS, `${runId}/expected.json`)
  const expected = JSON.parse(new TextDecoder().decode(raw)) as ExpectedJson
  const sample = await resolveSampleFile(expected.inputValue)
  if (!sample) return null

  const bundleLocal = {
    schemaVersion: 1 as const,
    sampleFile: `sample/${sample.filename}`,
    outputColumn: expected.outputColumn,
    outputType: expected.outputType,
    predictions: expected.predictions,
  }
  return { expectedJson: strToU8(JSON.stringify(bundleLocal, null, 2)), sample }
}

/** A Dart package/app name must be lowercase_with_underscores and start with a letter — pubspec.yaml's `name:` rejects anything else. */
function dartPackageName(name: string): string {
  const snake = name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '')
  const safe = /^[a-z]/.test(snake) ? snake : `app_${snake}`
  return safe || 'theseus_app'
}

type ZipInput = Parameters<typeof zipSync>[0]

async function assembleFiles(exportId: string): Promise<ZipInput> {
  const row = await db.query.modelExports.findFirst({
    where: { id: exportId },
    with: { run: { with: { project: true } } },
  })
  if (!row) throw new Error(`Export ${exportId} not found`)
  const { run, tier, format, lang } = row
  if (!run.project) throw new Error(`Run ${run.id}'s project not found`)

  const task = getTaskDescriptor(run.project.task)
  const modelBytes = await downloadFile(CONSTANTS.BUCKET_MODELS, exportKey(run.id, format as ExportFormat))
  const preprocessing = await extractPreprocessing(run.id)
  const preprocessingJson = strToU8(JSON.stringify(preprocessing, null, 2))
  const labelsTxt = preprocessing.outputs.find((o) => o.classes)?.classes?.join('\n')

  const files: ZipInput = {}

  /** Places model.<format> / preprocessing.json / labels.txt under `prefix` (root for most langs, `assets/` for Flutter). */
  function placeModelFiles(prefix: string): void {
    files[`${prefix}model.${format}`] = [modelBytes, { level: 0 }]
    files[`${prefix}preprocessing.json`] = preprocessingJson
    if (labelsTxt) files[`${prefix}labels.txt`] = strToU8(labelsTxt)
  }

  if (tier === 'model') {
    placeModelFiles('')
    files['README.md'] = strToU8(
      renderReadme(
        { runName: run.name, taskLabel: task.label, format: format as ExportFormat, tier, hasVerify: false },
        null,
      ),
    )
    return files
  }

  // devkit / app: model tier's files, plus a generated client.
  const golden = await loadGoldenSample(run.id)

  /** Places expected.json / sample/<file> under `prefix`, mirroring placeModelFiles. No-op if there's no golden sample. */
  function placeGoldenSample(prefix: string): void {
    if (!golden) return
    files[`${prefix}expected.json`] = golden.expectedJson
    files[`${prefix}sample/${golden.sample.filename}`] = golden.sample.bytes
  }

  const vars = { RUN_NAME: run.name, TASK: task.label }
  const readmeVars = {
    runName: run.name,
    taskLabel: task.label,
    format: format as ExportFormat,
    tier,
    hasVerify: !!golden,
  }

  if (lang === 'python') {
    placeModelFiles('')
    placeGoldenSample('')
    files['theseus_client.py'] = strToU8(templates.python.client)
    files['example.py'] = strToU8(templates.python.example)
    if (golden) files['verify.py'] = strToU8(templates.python.verify)
    files['README.md'] = strToU8(renderReadme(readmeVars, 'python'))
  } else if (lang === 'typescript') {
    placeModelFiles('')
    placeGoldenSample('')
    files['client.ts'] = strToU8(templates.typescript.client)
    files['example.ts'] = strToU8(templates.typescript.example)
    if (golden) files['verify.ts'] = strToU8(templates.typescript.verify)
    files['README.md'] = strToU8(renderReadme(readmeVars, 'typescript'))
  } else if (lang === 'csharp') {
    placeModelFiles('')
    placeGoldenSample('')
    files['TheseusClient.cs'] = strToU8(templates.csharp.client)
    files['Program.cs'] = strToU8(templates.csharp.program)
    files['README.md'] = strToU8(renderReadme(readmeVars, 'csharp'))
  } else if (lang === 'java') {
    placeModelFiles('')
    placeGoldenSample('')
    files['TheseusClient.java'] = strToU8(templates.java.client)
    files['Main.java'] = strToU8(templates.java.main)
    files['README.md'] = strToU8(renderReadme(readmeVars, 'java'))
  } else if (lang === 'pwa') {
    // Fetched relative to index.html at runtime — root placement is correct.
    placeModelFiles('')
    placeGoldenSample('')
    files['index.html'] = strToU8(render(templates.pwa.indexHtml, vars))
    files['app.js'] = strToU8(templates.pwa.appJs)
    files['sw.js'] = strToU8(templates.pwa.serviceWorker)
    files['manifest.webmanifest'] = strToU8(render(templates.pwa.manifest, vars))
    files['icon.svg'] = strToU8(templates.pwa.icon)
    files['style.css'] = strToU8(templates.pwa.style)
    files['README.md'] = strToU8(renderReadme(readmeVars, 'pwa'))
  } else if (lang === 'flutter') {
    // Flutter only reads files declared as pubspec assets — assets/ prefix required.
    placeModelFiles('assets/')
    placeGoldenSample('assets/')
    files['pubspec.yaml'] = strToU8(
      render(templates.flutter.pubspec, { ...vars, PACKAGE_NAME: dartPackageName(run.name) }),
    )
    files['lib/main.dart'] = strToU8(render(templates.flutter.main, vars))
    files['lib/theseus_client.dart'] = strToU8(templates.flutter.client)
    files['README.md'] = strToU8(renderReadme(readmeVars, 'flutter'))
  } else {
    throw new Error(`Export ${exportId}: tier '${tier}' requires a lang, got none`)
  }

  return files
}

/** Build one export's zip, upload it, and flip the row to ready/failed. Never throws — failures are recorded on the row. */
export async function buildBundle(exportId: string): Promise<void> {
  return record('export.assemble', async (span) => {
    span.setAttribute('theseus.export_id', exportId)
    try {
      const files = await assembleFiles(exportId)
      const zipped = zipSync(files, { level: 6 })

      const row = await db.query.modelExports.findFirst({ where: { id: exportId } })
      if (!row) throw new Error(`Export ${exportId} not found`)

      const key = bundleKey(row.runId, exportId)
      await uploadFile(CONSTANTS.BUCKET_MODELS, key, zipped, 'application/zip')
      const checksum = new Bun.CryptoHasher('sha256').update(zipped).digest('hex')

      await db
        .update(modelExports)
        .set({ status: 'ready', bundleKey: key, byteSize: zipped.byteLength, checksum, readyAt: new Date() })
        .where(eq(modelExports.id, exportId))
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e)
      console.error(`[export] assembly failed for ${exportId}:`, e)
      await db
        .update(modelExports)
        .set({ status: 'failed', failedMessage: message })
        .where(eq(modelExports.id, exportId))
    }
  })
}

File: ai_service/theseus/export/bundle.py
"""Assemble an export bundle zip and upload it (ported from server/lib/export/bundle.ts).

assemble_files() is pure (bytes in, file list out) and delegates the layout to the export format
class (theseus/export/formats/), so each format's layout can be tested on its own. build_bundle() is
the thin async wrapper that loads the inputs, zips in the export executor, uploads, and owns the
final `ready` / `failed` transition. It never raises.
"""

import asyncio
import hashlib
import io
import json
import logging
import uuid
import zipfile
from typing import Any

import sqlalchemy as sa

from theseus import constants as C
from theseus.backends.base import ARTIFACT_FILENAMES
from theseus.db.base import get_sessionmaker
from theseus.db.models import ModelExport, Project, TrainingRun
from theseus.export.common import TEMPLATES, BundleFile, GoldenSample, dart_package_name, dumps, render, template
from theseus.export.formats.base import BundleContext
from theseus.export.metadata import extract_preprocessing
from theseus.export.registry import get_export_format
from theseus.jobs.executors import export_executor, run_in_executor
from theseus.services import storage
from theseus.services.task_registry import get_task_descriptor

logger = logging.getLogger(__name__)

# A fixed timestamp keeps a bundle reproducible: the same inputs always produce the same zip bytes.
ZIP_EPOCH = (2020, 1, 1, 0, 0, 0)

__all__ = ["TEMPLATES", "BundleFile", "GoldenSample", "dart_package_name", "dumps", "render", "template"]


def resolve_sample_file(input_value: Any, download) -> tuple[str, bytes] | None:
    """Turn expected.json's raw inputValue into an actual file to embed under sample/.

    An s3:// URI (file-backed modalities) is downloaded; a string becomes input.txt; anything else
    becomes input.json. download(bucket, key) may raise, which yields None (no sample, so no verify).
    """
    if isinstance(input_value, str) and input_value.startswith("s3://"):
        rest = input_value[len("s3://") :]
        if "/" not in rest:
            return None
        bucket, key = rest.split("/", 1)
        ext = key[key.rfind(".") :] if "." in key else ""
        try:
            return f"input{ext}", download(bucket, key)
        except Exception:
            return None
    if isinstance(input_value, str):
        return "input.txt", input_value.encode()
    if input_value is not None:
        return "input.json", dumps(input_value).encode()
    return None


def to_bundle_local(expected: dict[str, Any], sample_filename: str) -> bytes:
    """expected.json rewritten to reference the sample by its bundle-local path."""
    return dumps(
        {
            "schemaVersion": 1,
            "sampleFile": f"sample/{sample_filename}",
            "outputColumn": expected.get("outputColumn"),
            "outputType": expected.get("outputType"),
            "predictions": expected.get("predictions"),
        }
    ).encode()


def assemble_files(
    *,
    run_name: str,
    task_label: str,
    format_id: str,
    model_bytes: bytes,
    preprocessing: dict[str, Any],
    golden: GoldenSample | None,
) -> list[BundleFile]:
    fmt = get_export_format(format_id)
    ctx = BundleContext(
        run_name=run_name,
        task_label=task_label,
        format_label=fmt.label,
        artifact_filename=ARTIFACT_FILENAMES[fmt.artifact],
        model_bytes=model_bytes,
        preprocessing=preprocessing,
        golden=golden if fmt.needs_golden else None,
    )
    fmt.assemble(ctx)
    return ctx.files


def make_zip(files: list[BundleFile]) -> bytes:
    buf = io.BytesIO()
    with zipfile.ZipFile(buf, "w") as zf:
        for f in files:
            info = zipfile.ZipInfo(f.path, date_time=ZIP_EPOCH)
            info.external_attr = 0o644 << 16
            zf.writestr(
                info,
                f.data,
                compress_type=zipfile.ZIP_DEFLATED if f.compress else zipfile.ZIP_STORED,
                compresslevel=6,
            )
    return buf.getvalue()


# -- Orchestration ---------------------------------------------------------------------------


async def _load_golden(run_id: uuid.UUID) -> GoldenSample | None:
    """The worker-written expected.json (if any) and its resolved sample file, with bundle-local paths."""
    loop = asyncio.get_running_loop()
    key = storage.expected_sample_key(str(run_id))
    if not await loop.run_in_executor(None, storage.file_exists, C.BUCKET_MODELS, key):
        return None
    expected = json.loads(await loop.run_in_executor(None, storage.download_bytes, C.BUCKET_MODELS, key))
    sample = await loop.run_in_executor(None, resolve_sample_file, expected.get("inputValue"), storage.download_bytes)
    if sample is None:
        return None
    filename, data = sample
    return GoldenSample(to_bundle_local(expected, filename), filename, data)


async def _assemble(export_id: uuid.UUID) -> tuple[uuid.UUID, list[BundleFile]]:
    async with get_sessionmaker()() as session:
        row = await session.get(ModelExport, export_id)
        if row is None:
            raise ValueError(f"Export {export_id} not found")
        run = await session.get(TrainingRun, row.run_id)
        project = await session.get(Project, run.project_id) if run else None
        if run is None or project is None:
            raise ValueError(f"Run {row.run_id} or its project not found")
        preprocessing = await extract_preprocessing(session, run.id)
        format_id = row.format
        run_id, run_name = run.id, run.name
        task_label = get_task_descriptor(project.task).label

    fmt = get_export_format(format_id)
    model_bytes = await run_in_executor(
        None, storage.download_bytes, C.BUCKET_MODELS, storage.export_key(str(run_id), ARTIFACT_FILENAMES[fmt.artifact])
    )
    golden = await _load_golden(run_id) if fmt.needs_golden else None
    files = assemble_files(
        run_name=run_name, task_label=task_label, format_id=format_id,
        model_bytes=model_bytes, preprocessing=preprocessing, golden=golden,
    )  # fmt: skip
    return run_id, files


async def build_bundle(export_id: uuid.UUID) -> None:
    """Build one export zip, upload it, and flip assembling -> ready | failed. Never raises.

    Both transitions are guarded on status = 'assembling': if startup recovery or a lease reaper
    already moved the row on, this result is stale and must not overwrite it.
    """
    try:
        run_id, files = await _assemble(export_id)
        zipped = await run_in_executor(export_executor, make_zip, files)
        key = storage.bundle_key(str(run_id), str(export_id))
        await run_in_executor(None, storage.upload_bytes, C.BUCKET_MODELS, key, zipped, "application/zip")
        checksum = hashlib.sha256(zipped).hexdigest()
        values: dict[str, Any] = {
            "status": "ready", "bundle_key": key, "byte_size": len(zipped), "checksum": checksum,
            "ready_at": sa.func.now(), "claimed_by": None, "lease_expires_at": None,
        }  # fmt: skip
    except Exception as e:
        logger.exception("Export assembly failed for %s", export_id)
        values = {
            "status": "failed", "failed_message": str(e)[:500], "claimed_by": None, "lease_expires_at": None,
        }  # fmt: skip
    try:
        async with get_sessionmaker()() as session:
            await session.execute(
                sa.update(ModelExport)
                .where(ModelExport.id == export_id, ModelExport.status == "assembling")
                .values(**values)
            )
            await session.commit()
    except Exception:
        logger.exception("Could not record the assembly outcome for export %s", export_id)
